File: src/util/access/V.java
package util.access;

import java.util.function.BiConsumer;
import java.util.function.Consumer;
import javafx.beans.InvalidationListener;
import javafx.beans.property.SimpleObjectProperty;
import javafx.beans.value.ChangeListener;
import org.reactfx.Subscription;

/**
 * Var/variable - simple object wrapper similar to {@link javafx.beans.property.Property}, but
 * simpler (no binding) and with the ability to apply value change.
 * <p/>
 * Does not permit null values.
 *
 * @author Martin Polakovic
 */
public class V<T> extends SimpleObjectProperty<T> implements ApplicableValue<T> {

	private Consumer<T> applier;

	public V(T val) {
		setValue(val);
	}

	public V(T val, Consumer<T> applier) {
		this(val);
		setApplier(applier);
	}

	public V(T val, Runnable applier) {
		this(val, v -> applier.run());
	}

	/** {@inheritDoc} */
	@Override
	public void applyValue(T val) {
		if (applier!=null) applier.accept(val);
	}

	/**
	 * Sets applier. Applier is a code that applies the value in any way.
	 *
	 * @param applier or null to disable applying
	 */
	public final void setApplier(Consumer<T> applier) {
		this.applier = applier;
	}

	/**
	 * Gets applier. Applier is a code that applies the value. It can do anything.
	 * Default null.
	 *
	 * @return applier or null if none.
	 */
	public Consumer<T> getApplier() {
		return applier;
	}

	public Subscription onChange(Consumer<? super T> action) {
		ChangeListener<T> l = (o, ov, nv) -> action.accept(nv);
		addListener(l);
		return () -> removeListener(l);
	}

	public Subscription onChange(BiConsumer<? super T,? super T> action) {
		ChangeListener<T> l = (o, ov, nv) -> action.accept(ov, nv);
		addListener(l);
		return () -> removeListener(l);
	}

	public Subscription maintain(Consumer<? super T> action) {
		ChangeListener<T> l = (o, ov, nv) -> action.accept(nv);
		addListener(l);
		action.accept(getValue());
		return () -> removeListener(l);
	}

	public Subscription onInvalid(Runnable action) {
		InvalidationListener l = o -> action.run();
		addListener(l);
		return () -> removeListener(l);
	}

}

File: src/util/access/V.kt
package util.access

import javafx.beans.property.SimpleObjectProperty
import org.reactfx.Subscription
import util.functional.invoke
import util.reactive.attach
import util.reactive.changes
import java.util.function.BiConsumer
import java.util.function.Consumer

fun <T> v(value: T) = V(value)
fun <T> v(value: T, onChange: (T) -> Unit) = V(value, Consumer { onChange(it) } )
fun <T> vn(value: T? = null) = V(value)
fun <T> vn(value: T?, onChange: (T?) -> Unit) = V(value, Consumer { onChange(it) } )

/**
 * Var/variable - simple object wrapper similar to [javafx.beans.property.Property], but
 * simpler (no binding) and with the ability to apply value change.
 *
 * Does not permit null values.
 */
open class V<T> : SimpleObjectProperty<T>, ApplicableValue<T> {

    var applier: Consumer<T>

    /**
     * Sets applier. Applier is a code that applies the value in any way.
     *
     * @param applier or null to disable applying
     */
    @JvmOverloads
    constructor(value: T, applier: Consumer<T> = Consumer<T> {}) {
        this.value = value
        this.applier = applier
    }

    constructor(value: T, applier: Runnable) : this(value, Consumer { applier() })

    override fun applyValue(`val`: T) {
        applier(`val`)
    }

    fun onChange(action: Consumer<in T>) = attach { action(it) }

    fun onChange(action: BiConsumer<in T, in T>) = changes { ov, nv -> action(ov, nv) }

    fun maintain(action: Consumer<in T>): Subscription {
        action(value)
        return attach { action(it) }
    }

    fun onInvalid(action: Runnable) = attach { action() }

}

open class VNullable<T>: V<T?> {

    @JvmOverloads
    constructor(value: T?, applier: Consumer<T?> = Consumer {}) : super(value, applier)

}